File: solutions/java/collatz-conjecture/1/src/main/java/CollatzCalculator.java
import java.util.stream.LongStream;

class CollatzCalculator {

	int computeStepCount(int start) {
		if (start <= 0) {
			throw new IllegalArgumentException("Only positive integers are allowed");
		}

		return (int) LongStream.iterate(start, this::isNotOne, this::nextCollatzValue)
				             .count();
	}

	private long nextCollatzValue(long n) {
		return (n % 2 == 0)
            ? n / 2
            : 3 * n + 1;
	}

	private boolean isNotOne(long n) {
		return n != 1;
	}
}

File: solutions/kotlin/collatz-conjecture/1/src/main/kotlin/CollatzCalculator.kt
object CollatzCalculator {
	fun computeStepCount(start: Int): Int {
		require(start > 0)
		return countSteps(start, 0)
	}

	private tailrec fun countSteps(current: Int, steps: Int): Int {
		return if (current == 1) {
			steps
		}
		else if (current % 2 == 0) {
			countSteps(current / 2, steps + 1)
		}
		else {
			countSteps(current * 3 + 1, steps + 1)
		}
	}
}